### ISSUE TITLE
Is similarity wrong ??

### ISSUE THREAD

**Issue by ramsanegl** [blingenf/copydetect#33]
Inside the `compare_files` method, we are using
```py
..
..
slices1 = get_copied_slices(idx1, file1_data.k)
.. 
token_overlap1 = np.sum(slices1[1] - slices1[0])
..
..
if len(file1_data.filtered_code) > 0:
        similarity1 = token_overlap1 / file1_data.token_coverage
..
```
to find the similarity.

But here, I think slices are the actual character(not tokens) slices. When we find the similarity, shouldn't we be divicing using `len(file_data.raw_code)` instead ? (Assuming token coverage is actually the number of tokens (not characters) in the code file)

**Comment by ramsanegl** [comment]
Am I missing something here ??

Cause when I compared exact same file, token coverage and overlap are same so similarities are 100%. But when I compared with slighlty modified files, then when I took the actual hashes that got matched from `find_fingerprint_overlap` method and divide with actual number of hashes in the file, I'm getting different score.
```py
..
copied_hashes, token_overlap, (sim1, sim2), (slices1, slices2) = compare_files(test_file, ref_file))
..
..
score = len(copied_hashes)/len(self.detector.file_data[test_f].hashes)
print(score, sim1)
..
..
```

here, in compare_files, I modified the `find_fingerprint_overlap` method to return the actual hashes that got matched

```py
def find_fingerprint_overlap(hashes1, hashes2, idx1, idx2):
    """Finds the indexes of overlapping values between two lists of
    hashes. Returns two lists of indexes, one for the first hash list
    and one for the second. The indexes of the original hashes are
    provided in case boilerplate results in gaps.
    """
    overlap, ol_idx1, ol_idx2 = np.intersect1d(hashes1, hashes2,
        return_indices=True, assume_unique=True)
    return ol_idx1, idx1[ol_idx1], idx2[ol_idx2]
```

**Comment by ramsanegl** [comment]
This I found out when I was trying to get the overall matched code for a single test_file when compared with all the ref_files.

**Comment by blingenf** [comment]
The detection loop works over k-grams, so the tokenized code is stitched back together into a string after tokenization. A file post-tokenization might look like this, for example:

`defF(V,V):SV=[hash(V[V:V+V])forVinrange(len(V)-V+1)]returnV.V(V)`

It's been a little while since I've looked at this code but I believe both the token coverage and the slice indices refer to this string, not the raw code. So when I say "tokens", I mean characters in the tokenized string.

Hash overlap divided by number of hashes is actually how copydetect 0.4.0 and 0.4.1 computed similarity. The problem with this method is that hash overlap doesn't always map very well back to how similar the original files are: there's an example of this issue in #19.

**Comment by blingenf** [comment]
The short answer is that the use case you're describing just isn't possible with how the code is structured.

The detection code was designed around pairwise similarity, so pretty much all the functions assume comparison between two files. I think the easiest way to implement a one-to many comparison would look something like this (adapted from `compare_files` in `detector.py`):

```python
slice_map = {}
similarity_statistics = {}
for file2_data in reference_files:
    # find any overlapping hashes
    overlap, ol_idx1, ol_idx2 = np.intersect1d(
        file1_data.hashes, file2_data.hashes,
        return_indices=True, assume_unique=True
    )
    if overlap.shape[0] > 0:
        # if there is overlap, store the copied slices under the name
        # of the copied file and remove the hashes from the fingerprint
        # object to avoid duplicate detection
        idx1 = file1_data.hash_idx[ol_idx1]
        slices = get_copied_slices(idx1, file1_data.k)
        slice_map[file2_data.filename] = slices
        file1_data.hashes = np.delete(file1_data.hashes, ol_idx1)

        # convert slices computed for the tokenized code into slices
        # for the original/untokenized code
        untokenized_slices = slices.copy()
        if len(file1_data.offsets) > 0:
            untokenized_slices += file1_data.offsets[:, 1][
                np.clip(np.searchsorted(file1_data.offsets[:, 0], untokenized_slices),
                0, file1_data.offsets.shape[0] - 1)
            ]

        # all three possible similarity metrics
        similarity_statistics[file2_data.filename] = {
            # number of fingerprints (n-grams) copied:
            "hashes": idx1.shape[0],
            # number of tokenized characters copied:
            "tokens": np.subtract(slices[1], slices[0]).sum(),
            # number of total characters copied (not considering tokenization):
            "characters": np.subtract(untokenized_slices[1], untokenized_slices[0]).sum()
        }
```
To compute similarity scores the first metric would be divided by fingerprint count (`file1_data.hashes`), the second by tokenized string length (`len(file1_data.filtered_code)`), and the third by original string length (`len(file1_data.raw_code)`).

Ultimately, however, this is just a thrown-together solution (it's not particularly efficient and modifying the hashes in place is not a good solution). A proper solution would unfortunately require a pretty comprehensive refactor of the current code.

**Comment by ramsane** [comment]
True. Even if we manage to do that. There is one more case to handle. As hashes are simply a set of hashes, not a list, if there is any repeated fingerprint (repeated k-gram string), that got copied, we should not remove the option to compare the same thing elsewhere. It is a different problem altogether.  Even If we decide to solve it, as you said, it would be like writing the entire library all over again.

Anyway, Thank you for your response. I greatly appreciate the help. 😊

**Comment by blingenf** [comment]
Closing this as the issue for matching similarity across multiple files is open in #44 and the handling of duplicate hashes was corrected by #48
